### PR TITLE
Use AddOrUpdate() to register runspace

### PR DESCRIPTION
**Pull request by iSazonov**
# PR Summary

Simplify code to register a runspace by using AddOrUpdate().
Reduce locks that is good for performance in multi-runspace environment.

## PR Context

See #2837

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)

### CODE REVIEW

File: src/System.Management.Automation/engine/runtime/Operations/ClassOps.cs [hunk 1]
```diff
@@ -42,26 +42,19 @@ internal SessionStateKeeper()
...
-                _stateMap.Add(rsToUse, ssToUse);
-            }
+
+            // If the key exists but the corresponding value is not what we should use, then update the key/value pair.
+            // This could happen when a powershell class is defined in a module and the module gets reloaded.
+            // In such case, the same TypeDefinitionAst instance will get reused,
+            // but should be associated with the SessionState from the new module, instead of the one from the old module.
             // If the key exists and the corresponding value is the one we should use, then do nothing.
+
+            _stateMap.AddOrUpdate(runspaceToUse, sessionStateToUse);
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/runtime/Operations/ClassOps.cs]: Update is not need for most of the cases, but `AddOrUpdate` will always update the value, even though the existing value is already what we want.

Comment (daxian-dbw) [inline src/System.Management.Automation/engine/runtime/Operations/ClassOps.cs]: This change is not right. Update is not need for most of the cases, but `AddOrUpdate` will always update the value, even though the existing value is already what we want.

**Review by daxian-dbw — CHANGES_REQUESTED** [review]

File: src/System.Management.Automation/engine/runtime/Operations/ClassOps.cs (hunk 1 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/runtime/Operations/ClassOps.cs]: > for most of the cases

My mistake is that I skipped that AddOrUpdate does a lock. 
Update itself is not expensive for references.
I think we could only replace Remove()+ Add() (lines 61-62 in original code) with single lock call AddOrUpdate() - this is twice faster (both Remove() and Add() do lock and FindEntry) and exclude a race condition.

Comment (daxian-dbw) [reply src/System.Management.Automation/engine/runtime/Operations/ClassOps.cs]: > replace Remove()+ Add() (lines 61-62 in original code) with single lock call AddOrUpdate()

This sounds about right. Please go ahead making the change.
Also, maybe change `!ssInMap.Equals(ssToUse)` to `ssInMap != ssToUse`

Comment (iSazonov) [reply src/System.Management.Automation/engine/runtime/Operations/ClassOps.cs]: Done.

**Review by daxian-dbw — APPROVED** [review]

**Pull request merged**